### PR TITLE
no crashing in rosserial python

### PR DESCRIPTION
**Pull request by gwendolynbrook**
Prevent complete crash in rosserial.

### CODE REVIEW

File: rosserial_python/src/rosserial_python/SerialClient.py [hunk 1]
```diff
@@ -134,7 +135,10 @@ def callback(self, msg):
         """ Forward message to serial device. """
         data_buffer = StringIO.StringIO()
         msg.serialize(data_buffer)
-        self.parent.send(self.id, data_buffer.getvalue())
+        try:
+            self.parent.send(self.id, data_buffer.getvalue())
+        except SerialException as e:
+            rospy.logwarn("Send failed. Port might be missing {0}".format(e.what()))
```
Comment (raghavakumar8) [inline rosserial_python/src/rosserial_python/SerialClient.py]: `logwarn_throttle`?

File: rosserial_python/src/rosserial_python/SerialClient.py [hunk 2]
```diff
@@ -365,33 +371,45 @@ def __init__(self, port=None, baud=57600, timeout=5.0):
 
         self.buffer_out = -1
         self.buffer_in = -1
+        self.resetCallbacks()
+
+        rospy.sleep(2.0) # TODO
```
Comment (raghavakumar8) [inline rosserial_python/src/rosserial_python/SerialClient.py]: What happens if you get rid of this delay?

File: rosserial_python/src/rosserial_python/SerialClient.py [hunk 3]
```diff
@@ -520,7 +538,8 @@ def run(self):
             except IOError:
                 # One of the read calls had an issue. Just to be safe,
                 # request that the client reinitialize their topics.
-                rospy.logwarn("Requesting reinitialization of client topics!")
+                # rospy.logwarn("Requesting reinitialization of client topics!")
+                time.sleep(0.002)
```
Comment (raghavakumar8) [inline rosserial_python/src/rosserial_python/SerialClient.py]: What does this delay do?

File: rosserial_python/src/rosserial_python/SerialClient.py [hunk 4]
```diff
@@ -770,8 +789,12 @@ def send(self, topic, msg):
                     msg_checksum = 255 - ( ((topic&255) + (topic>>8) + sum([ord(x) for x in msg]))%256 )
                     data = "\xff" + self.protocol_ver  + chr(length&255) + chr(length>>8) + chr(msg_len_checksum) + chr(topic&255) + chr(topic>>8)
                     data = data + msg + chr(msg_checksum)
-                    self.port.write(data)
-                    return length
+                    try:
+                        self.port.write(data)
+                        return length
+                    except SerialException as e:
+                        rospy.logwarn("Send failed")
```
Comment (raghavakumar8) [inline rosserial_python/src/rosserial_python/SerialClient.py]: `logwarn_throttle`?

**Review by raghavakumar8 — COMMENTED** [review]
Did you measure recovery time before/after this change? Would be nice to see some numbers!

File: rosserial_python/src/rosserial_python/SerialClient.py (hunk 2 above)
Comment (gwendolynbrook) [reply rosserial_python/src/rosserial_python/SerialClient.py]: Unclear .... will test what removing it does when getting other numbers.

Comment (gwendolynbrook) [reply rosserial_python/src/rosserial_python/SerialClient.py]: unclear -- it was already there (previous line 384;) can test affect of removing it.

File: rosserial_python/src/rosserial_python/SerialClient.py (hunk 3 above)
Comment (gwendolynbrook) [reply rosserial_python/src/rosserial_python/SerialClient.py]: hmmm..... not particularly useful; bit of time for the port to come back up?

**Pull request merged**